cmd/showalloc/showalloc.go

```diff
@@ -0,0 +1,201 @@
+/*
+© 2020 Nokia
+Licensed under the BSD 3-Clause License
+SPDX-License-Identifier: BSD-3-Clause
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+
+	danmtypes "github.com/nokia/danm/crd/apis/danm/v1"
+	danmclientset "github.com/nokia/danm/crd/client/clientset/versioned"
+	"github.com/nokia/danm/pkg/bitarray"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func addIP(ip net.IP, num uint32) net.IP { // last 4 bytes are considered only
+	oldIP := ip.To16()
+	newIP := make(net.IP, len(oldIP))
+	copy(newIP, oldIP)
+	v := uint32(oldIP[12])<<24 + uint32(oldIP[13])<<16 + uint32(oldIP[14])<<8 + uint32(oldIP[15])
+	v += num
+	newIP[15] = byte(v & 0xFF)
+	newIP[14] = byte((v >> 8) & 0xFF)
+	newIP[13] = byte((v >> 16) & 0xFF)
+	newIP[12] = byte((v >> 24) & 0xFF)
+	return newIP
+}
+
+func diffIP(ip1, ip2 net.IP) uint32 { // last 4 bytes are considered only
+	i1 := ip1.To16()
+	i2 := ip2.To16()
+	u1 := uint32(i1[12])<<24 + uint32(i1[13])<<16 + uint32(i1[14])<<8 + uint32(i1[15])
+	u2 := uint32(i2[12])<<24 + uint32(i2[13])<<16 + uint32(i2[14])<<8 + uint32(i2[15])
+	return uint32(u2) - uint32(u1)
+}
+
+func findEpByIP(epList *danmtypes.DanmEpList, ip net.IP, v6 bool) *danmtypes.DanmEp {
+	var epIP net.IP
+	for _, ep := range epList.Items {
+		if v6 {
+			epIP, _, _ = net.ParseCIDR(ep.Spec.Iface.AddressIPv6) // safe to ignore err as it is a validated property
+		} else {
+			epIP, _, _ = net.ParseCIDR(ep.Spec.Iface.Address) // safe to ignore err as it is a validated property
+		}
+		if epIP.Equal(ip) {
+			return &ep
+		}
+	}
+	return nil
+}
+
+func main() {
+	var (
+		// flags
+		kubeConfig, dnet, cnet, tnet, namespace string
+		v6, showAll                             bool
+		showCidrStr                             string
+		// for showCIDR feature
+		isShowCidrRequested bool = false
+		showCidr            *net.IPNet
+		// network parameters
+		spec                                  danmtypes.DanmNetSpec
+		kind, alloc, cidr, poolStart, poolEnd string
+		// others
+		i, iStart, iEnd uint32
+		status          string
+		err             error
+	)
+
+	flag.StringVar(&kubeConfig, "kubeconfig", "", "Absolute path to a valid kubeconfig file. Only required if ShowAlloc runs out-of-cluster.")
```

Comment:
considering this is a command line utility requiring RBAC kubeconfig is mandatory, right?
also the RBAC profile required to function would be nice to have

Reply:
Right, `integration/manifests/showalloc/showalloc.yaml` is added.

I expect that this binary will be part of the `danm-utils` container. The easiest way to run this `showalloc` is to execute it inside the Cleaner Pod (as it is continuously running). Cleaner's RBAC/serviceaccount is adequate for `showalloc`.

Reply:
hmm, that's one way to go about it. I was thinking more about a standalone utility, but can do both of course